druntime/src/core/internal/array/appending.d

```diff
@@ -22,25 +26,84 @@ private enum isCopyingNothrow(T) = __traits(compiles, (ref T rhs) nothrow { T lh
  *  n = how many new elements to extend it with
  * Returns:
  *  The new value of `px`
- * Bugs:
- *  This function template was ported from a much older runtime hook that bypassed safety,
- *  purity, and throwabilty checks. To prevent breaking existing code, this function template
- *  is temporarily declared `@trusted pure` until the implementation can be brought up to modern D expectations.
  */
 ref Tarr _d_arrayappendcTX(Tarr : T[], T)(return ref scope Tarr px, size_t n) @trusted
 {
-    // needed for CTFE: https://github.com/dlang/druntime/pull/3870#issuecomment-1178800718
+    import core.internal.traits: Unqual;
+
+    alias Unqual_T = Unqual!T;
+    alias Unqual_Tarr = Unqual_T[];
+    enum isshared = is(T == shared);
+    auto unqual_px = cast(Unqual_Tarr) px;
+
+    // Ignoring additional attributes allows reusing the same generated code
+    px = cast(Tarr)_d_arrayappendcTX_(unqual_px, n, isshared);
+    return px;
+}
+
+private ref Tarr _d_arrayappendcTX_(Tarr : T[], T)(return ref scope Tarr px, size_t n, bool isshared) @trusted
+{
     version (DigitalMars) pragma(inline, false);
     version (D_TypeInfo)
     {
-        auto ti = typeid(Tarr);
+        // Short circuit if no data is being appended.
+        if (n == 0)
+            return px;
+
+        import core.stdc.string : memcpy, memset;
+        import core.internal.lifetime : __doPostblit;
+        import core.internal.array.utils: __arrayAlloc, newCapacity, __typeAttrs;
+        import core.internal.gc.blockmeta : PAGESIZE;
+        import core.exception: onOutOfMemoryError;
+        import core.memory: GC;
+
+        alias BlkAttr = GC.BlkAttr;
 
-        // _d_arrayappendcTX takes the `px` as a ref byte[], but its length
-        // should still be the original length
-        auto pxx = (cast(byte*)px.ptr)[0 .. px.length];
-        ._d_arrayappendcTX(ti, pxx, n);
-        px = (cast(T*)pxx.ptr)[0 .. pxx.length];
+        enum sizeelem = T.sizeof;
+        auto length = px.length;
+        auto newlength = length + n;
+        auto newsize = newlength * sizeelem;
+        auto size = length * sizeelem;
 
+        if (!gc_expandArrayUsed((cast(void*)px.ptr)[0 .. size], newsize, isshared))
+        {
+            // could not set the size, we must reallocate.
+            auto newcap = newCapacity(newlength, sizeelem);
+            auto attrs = __typeAttrs!T(cast(void*)px.ptr) | BlkAttr.APPENDABLE;
+
+            // use this static enum to avoid recomputing TypeInfo for every call.
+            static enum ti = typeid(T);
+            void* ptr = GC.malloc(newcap, attrs, ti);
```

Comment:
```suggestion
            void* ptr = GC.malloc(newcap, attrs, typeid(T));
```
`typeid(Type)` is a static constant, we do not need to recalculate. It's only a runtime call for class instances.